prime-router/src/main/kotlin/azure/BlobAccess.kt

```diff
@@ -31,10 +31,41 @@ class BlobAccess(
         val digest: ByteArray,
     )
 
-    fun uploadBody(report: Report): BlobInfo {
+    /**
+     * Upload the [report] to the blob store using the [action] to determine a folder as needed.  A [subfolderName]
+     * is optional and is added as a prefix to the blob filename.
+     * @return the information about the uploaded blob
+     */
+    fun uploadBody(
+        report: Report,
+        subfolderName: String? = null,
+        action: Event.EventAction = Event.EventAction.NONE
+    ): BlobInfo {
         val (bodyFormat, blobBytes) = createBodyBytes(report)
+        return uploadBody(bodyFormat, blobBytes, report.name, subfolderName, action)
+    }
+
+    /**
+     * Upload a raw [blobBytes] in the [bodyFormat] for a given [reportName].  The [action] is used to determine
+     * the folder to store the blob in.  A [subfolderName] name is optional.
+     * @return the information about the uploaded blob
+     */
+    fun uploadBody(
+        bodyFormat: Report.Format,
+        blobBytes: ByteArray,
+        reportName: String,
+        subfolderName: String? = null,
+        action: Event.EventAction = Event.EventAction.NONE
+    ): BlobInfo {
+        val subfolderNameChecked = if (subfolderName.isNullOrBlank()) "" else "$subfolderName/"
+        val blobName = when (action) {
+            Event.EventAction.RECEIVE -> "receive/$subfolderNameChecked$reportName"
+            Event.EventAction.SEND -> "ready/$subfolderNameChecked$reportName"
+            Event.EventAction.BATCH -> "batch/$subfolderNameChecked$reportName"
+            else -> error("Cannot determine folder to store blob.  Unsupported action $action for report $reportName")
```

Comment:
Carlos, this will only fix half the issue. We have receivers with the same file name format (Santa Clara County and California for example). If a file came in with results that would route to both, which is likely since Santa Clara County is in California, you'd still end up with a name collision when both files were written to these folders.

Reply:
@mauricereeves I think this is resolved here.  Let's make sure.
If I understand correctly, you can have the same filename be sent to two different receivers.  Since the folder structure is now ready/<receiver>/filename then the collision will not happen.  Is this correct?  That subfolderName gets passed the sender or receiver name, so it is added as a folder level.

Reply:
Yeah, I think we're ok as long as filenames are unique within a single receiver full name, since those are used as the subfolder names.